src/Webpage/Traits/HasParent.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+namespace Maxfactor\Support\Webpage\Traits;
+
+use Illuminate\Database\Eloquent\Builder;
+
+/**
+ * Allow any model to have nested Parents. This will handle the validating and generating of
+ * full nested slug paths.
+ *
+ * The model must have a `parent()` method.
+ *
+ * To filter specific folders to be removed when working with sub-domains, populate a protected
+ * array attribute on the model named `$domainMappedFolders`.
+ */
+trait HasParent
+{
+    /**
+     * The accessors to append to the model's array form.
+     *
+     * @var array
+     */
+    protected $appendsParentFields = [
+        'full_path',
+    ];
+
+    /**
+     * Called by constructor to load appends fields.
+     */
+    public function initHasParent()
+    {
+        $this->appends = array_merge($this->appends, $this->appendsParentFields);
+    }
+
+    /**
+     * Get full path.
+     *
+     * @return string Full path
+     */
+    protected function getFullPath($item = null)
+    {
+        if (!$item) {
+            $item = $this;
+        }
+
+        if (!$parent = $item->parent) {
+            return str_start($item->slug, '/');
+        }
+
+        return str_start(sprintf('%s%s', $this->getFullPath($parent), str_start($item->slug, '/')), '/');
+    }
+
+    /**
+     * Get root slug.
+     *
+     * @return string Root slug
+     */
+    public function getRootSlug()
+    {
+        $pathSections = explode('/', $this->getFullPath());
+
+        return collect($pathSections)->filter()->first();
+    }
+
+    /**
+     * Add full path attribute.
+     *
+     * @return string Full path
+     */
+    public function getFullPathAttribute()
+    {
+        $pathSections = explode('/', $this->getFullPath());
+
+        $slugsToExclude = array_wrap($this->domainMappedFolders);
+
+        $finalSlug = collect($pathSections)
+            ->filter()
+            ->reject(function ($slug) use ($slugsToExclude) {
+                return in_array($slug, $slugsToExclude);
+            })
+            ->implode('/');
+
+        return str_start($finalSlug, '/');
+    }
+
+    /**
+     * Lazy-load parent items
+     *
+     * @param Builder $query
+     * @return Builder
+     */
+    public function scopeWithParent(Builder $query)
+    {
+        return $query->with('parent');
+    }
+
+    /**
+     * Filter the model to only show items which match the full path
+     *
+     * @param Builder $query
+     * @param string $path
+     * @return Builder
+     */
+    public function scopeWhereFullPath(Builder $query, string $path)
+    {
+        $itemSlugs = explode('/', $path);
+
+        return $query->where('slug', '=', end($itemSlugs))
+            ->get()
```

Comment:
Just wrote this exact method on Leightons @m2de, however I added the `withParent()` local scope before `get()`.

What's your thoughts on this, as you're likely to want to eager load the `parent` relationship when scoping the query by an item's full path. I was thinking that you wouldn't want to call `withParent()` before `whereFullPath()` as otherwise you'd be needlessly loading all parents for items you're about to filter out.

Reply:
See line 92 ;)

Reply:
It doesn't execute the query until the end, so it won't eager load items which have been filtered out.